src/Microsoft.OData.Core/UriParser/ExpressionLexer.cs

```diff
@@ -447,6 +448,43 @@ internal string AdvanceThroughBalancedParentheticalExpression()
             return expressionText;
         }
 
+        /// <summary>
+        /// Get the current position in this lexer that can be used restore the lexer to this position later.
+        /// </summary>
+        /// <returns>
+        /// Returns a snapshot position used to call RestorePosition.
+        /// </returns>
+        internal ExpressionLexerPosition SnapshotPosition()
+        {
+            return new ExpressionLexerPosition(this, this.textPos, this.token);
+        }
+
+        /// <summary>
+        /// Sets the current position to the specified position.
+        /// </summary>
+        /// <param name="position">The position to restore, returned from SnapshotPosition.</param>
+        /// <remarks>
+        /// The specified position must have been retrieved by GetPostion on this instance.
+        /// </remarks>
+        internal void RestorePosition(ExpressionLexerPosition position)
+        {
+
+            if (position.Lexer != this)
+            {
+                throw new ArgumentException("Position was not taken from this ExpressionLexer instance.", "position");
+            }
+

```

Comment:
Do we really need to pass the lexer and do this validation? since this is all internal, this is really just there to assert *we* didn't do something wrong. We should minimally make this an assert (the error is meaningless to an enduser), if not remove it all together.